src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs

```diff
@@ -85,17 +70,30 @@ public override void VisitQueryExpression(QueryExpressionSyntax node)
 
         public override void VisitBinaryExpression(BinaryExpressionSyntax node)
         {
-            expressionsToVisit.Add(node.Right);
-            expressionsToVisit.Add(node.Left);
-        }
-        public override void VisitPostfixUnaryExpression(PostfixUnaryExpressionSyntax node)
-        {
-            expressionsToVisit.Add(node.Operand);
-        }
-        public override void VisitPrefixUnaryExpression(PrefixUnaryExpressionSyntax node)
-        {
-            expressionsToVisit.Add(node.Operand);
+            // The binary operators (except ??) are left-associative, and expressions of the form
+            // a + b + c + d .... are relatively common in machine-generated code. The parser can handle
+            // creating a deep-on-the-left syntax tree no problem, and then we promptly blow the stack during
+            // semantic analysis. Here we build an explicit stack to handle left recursion.
+
+            var operands = ArrayBuilder<ExpressionSyntax>.GetInstance();
+            ExpressionSyntax current = node;
+            do
+            {
+                var binOp = (BinaryExpressionSyntax)current;
+                operands.Push(binOp.Right);
+                current = binOp.Left;
+            }
+            while (current is BinaryExpressionSyntax);
```

Comment:
Is this to handle deeply nested patterns? Do we expect such cases? And are we testing this?